# EPC29 - ExcludeFromCodeCoverageAttribute should provide a message

This analyzer detects when `ExcludeFromCodeCoverageAttribute` is used without providing a justification message explaining why the code is excluded from coverage.

## Description

The analyzer warns when `ExcludeFromCodeCoverageAttribute` is applied without providing a `Justification` message. Including a justification helps other developers understand why certain code is excluded from coverage analysis.

## Code that triggers the analyzer

```csharp
using System.Diagnostics.CodeAnalysis;

public class Example
{
    // Bad: No justification provided
    [ExcludeFromCodeCoverage]
    public void SomeMethod()
    {
        // Why is this excluded? Unclear to other developers
    }
    
    // Also bad: Empty attribute
    [ExcludeFromCodeCoverage()]
    public class HelperClass
    {
        // No explanation for exclusion
    }
}
```

## How to fix

Provide a clear justification message:

```csharp
using System.Diagnostics.CodeAnalysis;

public class Example
{
    // Good: Clear justification provided
    [ExcludeFromCodeCoverage(Justification = "This method is only used for debugging and doesn't need test coverage")]
    public void DebugHelper()
    {
        // Debug-only code
    }
    
    [ExcludeFromCodeCoverage(Justification = "Generated code that doesn't require testing")]
    public class AutoGeneratedClass
    {
        // Auto-generated methods
    }
    
    [ExcludeFromCodeCoverage(Justification = "Platform-specific code covered by integration tests")]
    public void PlatformSpecificMethod()
    {
        // Platform-specific implementation
    }
    
    [ExcludeFromCodeCoverage(Justification = "Simple property with no logic to test")]
    public string SimpleProperty { get; set; }
}
```

## Common justifications

```csharp
// Generated code
[ExcludeFromCodeCoverage(Justification = "Auto-generated by code generator")]

// Defensive programming
[ExcludeFromCodeCoverage(Justification = "Defensive null check that should never be hit")]

// Platform-specific code
[ExcludeFromCodeCoverage(Justification = "Platform-specific code tested through integration tests")]

// Debug/development code
[ExcludeFromCodeCoverage(Justification = "Development helper method not used in production")]

// Exception handling
[ExcludeFromCodeCoverage(Justification = "Exception case difficult to reproduce in unit tests")]

// Legacy code
[ExcludeFromCodeCoverage(Justification = "Legacy code scheduled for removal in next version")]
```

## Benefits of providing justification

- Documents the reason for exclusion
- Helps during code reviews
- Makes it easier to evaluate whether exclusion is still appropriate
- Improves code maintainability
- Helps new team members understand the codebase
