# ERP042 - EventSource implementation is not correct

This analyzer detects various issues with EventSource implementations that can cause runtime errors or incorrect behavior.

## Description

The analyzer warns about incorrect EventSource implementations, including issues with event IDs, parameter mismatches, incorrect WriteEvent calls, and other problems that can cause ETW (Event Tracing for Windows) to fail at runtime.

## Code that triggers the analyzer

```csharp
using System.Diagnostics.Tracing;

[EventSource(Name = "MyEventSource")]
public sealed class ProblematicEventSource : EventSource
{
    public static readonly ProblematicEventSource Log = new ProblematicEventSource();
    
    // Bad: Duplicate event IDs
    [Event(1)]
    public void Event1() => WriteEvent(1); // ❌ ERP042
    
    [Event(1)] // ❌ ERP042 - Same ID as above!
    public void AnotherEvent() => WriteEvent(1);
    
    // Bad: Event ID mismatch
    [Event(2)]
    public void Event2(string message) => WriteEvent(3, message); // ❌ ERP042 - Wrong ID!
    
    // Bad: Parameter count mismatch
    [Event(3)]
    public void Event3(string msg, int count) => WriteEvent(3, msg); // ❌ ERP042 - Missing parameter!
    
    // Bad: Parameter order mismatch  
    [Event(4)]
    public void Event4(string first, int second) => WriteEvent(4, second, first); // ❌ ERP042 - Wrong order!
    
    // Bad: Unsupported parameter type
    [Event(5)]
    public void Event5(DateTime timestamp) => WriteEvent(5, timestamp); // ❌ ERP042 - DateTime not supported!
    
    // Bad: Missing parameter name in WriteEventCore
    [Event(6)]
    public void Event6(string message, int value) // ❌ ERP042
    {
        WriteEventCore(6, 1, new EventData[]
        {
            new EventData { DataPointer = /* message pointer */, Size = message.Length }
            // Missing second parameter!
        });
    }
}
```

## How to fix

Implement EventSource correctly:

```csharp
using System.Diagnostics.Tracing;

[EventSource(Name = "MyCompany-MyProduct-EventSource")]
public sealed class CorrectEventSource : EventSource
{
    public static readonly CorrectEventSource Log = new CorrectEventSource();
    private CorrectEventSource() { }
    
    // Good: Unique event IDs
    [Event(1, Level = EventLevel.Informational)]
    public void ApplicationStarted(string applicationName, string version) // ✅ Correct
    {
        if (IsEnabled())
        {
            WriteEvent(1, applicationName, version); // ✅ Correct - Matching ID and parameters
        }
    }
    
    // Good: Correct event ID and parameter mapping
    [Event(2, Level = EventLevel.Warning)]
    public void WarningOccurred(string component, string message, int errorCode) // ✅ Correct
    {
        if (IsEnabled(EventLevel.Warning))
        {
            WriteEvent(2, component, message, errorCode); // ✅ Correct - All parameters included
        }
    }
    
    // Good: Supported parameter types only
    [Event(3, Level = EventLevel.Error)]
    public void ErrorOccurred(string operation, int errorCode, long timestamp) // ✅ Correct
    {
        if (IsEnabled(EventLevel.Error))
        {
            WriteEvent(3, operation, errorCode, timestamp); // ✅ Correct
        }
    }
    
    // Good: Proper WriteEventCore usage
    [Event(4, Level = EventLevel.Verbose)]
    public void DetailedEvent(string category, string details) // ✅ Correct
    {
        if (IsEnabled(EventLevel.Verbose))
        {
            unsafe
            {
                fixed (char* categoryPtr = category)
                fixed (char* detailsPtr = details)
                {
                    WriteEventCore(4, 2, new EventData[] // ✅ Correct
                    {
                        new EventData 
                        { 
                            DataPointer = (IntPtr)categoryPtr, 
                            Size = (category.Length + 1) * sizeof(char) 
                        },
                        new EventData 
                        { 
                            DataPointer = (IntPtr)detailsPtr, 
                            Size = (details.Length + 1) * sizeof(char) 
                        }
                    });
                }
            }
        }
    }
    
    // Good: Using supported types and proper enum conversion
    [Event(5, Level = EventLevel.Informational)]
    public void StatusChanged(string component, int newStatus) // int instead of enum
    {
        if (IsEnabled())
        {
            WriteEvent(5, component, newStatus);
        }
    }
}
```

Handling complex types correctly:

```csharp
[EventSource(Name = "MyCompany-ComplexEvents")]
public sealed class ComplexEventSource : EventSource
{
    public static readonly ComplexEventSource Log = new ComplexEventSource();
    private ComplexEventSource() { }
    
    // Convert DateTime to supported type
    [Event(1)]
    public void TimestampEvent(string operation, DateTime timestamp)
    {
        if (IsEnabled())
        {
            // Convert DateTime to long (ticks) which is supported
            WriteEvent(1, operation, timestamp.Ticks);
        }
    }
    
    // Handle Guid by converting to string
    [Event(2)]
    public void CorrelationEvent(string operation, Guid correlationId)
    {
        if (IsEnabled())
        {
            WriteEvent(2, operation, correlationId.ToString());
        }
    }
    
    // Handle custom objects by extracting properties
    [Event(3)]
    public void UserActionEvent(string userId, string action, UserInfo userInfo)
    {
        if (IsEnabled())
        {
            // Extract supported properties from complex object
            WriteEvent(3, userId, action, userInfo.Name, userInfo.Age);
        }
    }
    
    // Overload for better usability while maintaining correctness
    public void UserActionEvent(string userId, string action, string userName, int userAge)
    {
        WriteEvent(3, userId, action, userName, userAge);
    }
}

public class UserInfo
{
    public string Name { get; set; }
    public int Age { get; set; }
}
```

Handling WriteEventWithRelatedActivityId correctly:

```csharp
[EventSource(Name = "MyCompany-ActivityEvents")]
public sealed class ActivityEventSource : EventSource
{
    public static readonly ActivityEventSource Log = new ActivityEventSource();
    private ActivityEventSource() { }
    
    // Good: First parameter must be Guid and named "relatedActivityId"
    [Event(1)]
    public void StartActivityWithRelated(Guid relatedActivityId, string activityName, string details)
    {
        if (IsEnabled())
        {
            WriteEventWithRelatedActivityId(1, relatedActivityId, activityName, details);
        }
    }
    
    // Good: Using proper parameter order and types
    [Event(2)]
    public void CompleteActivityWithRelated(Guid relatedActivityId, string activityName, long durationMs, bool success)
    {
        if (IsEnabled())
        {
            WriteEventWithRelatedActivityId(2, relatedActivityId, activityName, durationMs, success);
        }
    }
}
```

## Common EventSource rules

1. **Event IDs must be unique** within an EventSource
2. **Event ID in attribute must match WriteEvent call**
3. **Parameter count must match** between method signature and WriteEvent
4. **Parameter order must match** between method signature and WriteEvent
5. **Use only supported parameter types**:
   - Primitive types: `bool`, `byte`, `short`, `int`, `long`, `float`, `double`
   - `string`, `char`
   - `IntPtr`, `UIntPtr`
   - `Guid` (in some contexts)
6. **For WriteEventWithRelatedActivityId**: First parameter must be `Guid` named "relatedActivityId"
7. **Always check IsEnabled()** before expensive operations
8. **Use IsEnabled(level, keywords)** for conditional logging

## Supported parameter types

```csharp
// ✓ Supported types
public void GoodEvent(
    bool flag,
    byte b, short s, int i, long l,
    float f, double d,
    string text, char c,
    IntPtr ptr, UIntPtr uptr)
{
    WriteEvent(1, flag, b, s, i, l, f, d, text, c, ptr, uptr);
}

// ✗ Unsupported types (convert before logging)
public void BadEvent(
    DateTime dt,        // Convert to .Ticks (long)
    Guid guid,          // Convert to .ToString()
    MyEnum enumVal,     // Convert to (int)enumVal
    MyClass obj)        // Extract individual properties
{
    // Don't do this - will cause runtime errors
}
```
