# ERP031 - The API is not thread-safe

This analyzer detects usage of APIs that are not thread-safe in multi-threaded contexts.

## Description

The analyzer warns when APIs that are known to be not thread-safe are used in ways that could lead to race conditions or other thread-safety issues. This helps prevent subtle bugs that can occur in multi-threaded applications.

## Code that triggers the analyzer

```csharp
public class Example
{
    private List<string> _items = new List<string>(); // Not thread-safe
    
    public void AddItem(string item)
    {
        // Bad: List<T> is not thread-safe for concurrent modifications
        _items.Add(item); // Race condition possible
    }
    
    public void ProcessItems()
    {
        // Bad: Iteration while other threads might modify
        foreach (var item in _items) // Potential exception
        {
            ProcessItem(item);
        }
    }
}

public class SharedCounter
{
    private int _count = 0;
    
    public void Increment()
    {
        // Bad: Non-atomic read-modify-write operation
        _count++; // Race condition
    }
    
    public int GetCount()
    {
        return _count; // May return inconsistent value
    }
}
```

## How to fix

Use thread-safe alternatives or add proper synchronization:

```csharp
public class ThreadSafeExample
{
    // Option 1: Use thread-safe collection
    private ConcurrentBag<string> _items = new ConcurrentBag<string>();
    
    public void AddItem(string item)
    {
        // Good: ConcurrentBag is thread-safe
        _items.Add(item);
    }
    
    public void ProcessItems()
    {
        // Good: ConcurrentBag supports safe enumeration
        foreach (var item in _items)
        {
            ProcessItem(item);
        }
    }
}

public class ThreadSafeCounter
{
    private int _count = 0;
    
    public void Increment()
    {
        // Good: Atomic increment operation
        Interlocked.Increment(ref _count);
    }
    
    public int GetCount()
    {
        // Good: Atomic read
        return Interlocked.Read(ref _count);
    }
}
```

Using locks for synchronization:

```csharp
public class SynchronizedExample
{
    private readonly List<string> _items = new List<string>();
    private readonly object _lock = new object();
    
    public void AddItem(string item)
    {
        lock (_lock)
        {
            _items.Add(item);
        }
    }
    
    public void ProcessItems()
    {
        List<string> itemsCopy;
        
        // Create a copy under lock to avoid holding lock during processing
        lock (_lock)
        {
            itemsCopy = new List<string>(_items);
        }
        
        // Process outside of lock
        foreach (var item in itemsCopy)
        {
            ProcessItem(item);
        }
    }
    
    public int Count
    {
        get
        {
            lock (_lock)
            {
                return _items.Count;
            }
        }
    }
}
```

Using reader-writer locks for better performance:

```csharp
public class ReaderWriterExample
{
    private readonly List<string> _items = new List<string>();
    private readonly ReaderWriterLockSlim _lock = new ReaderWriterLockSlim();
    
    public void AddItem(string item)
    {
        _lock.EnterWriteLock();
        try
        {
            _items.Add(item);
        }
        finally
        {
            _lock.ExitWriteLock();
        }
    }
    
    public IEnumerable<string> GetItems()
    {
        _lock.EnterReadLock();
        try
        {
            return _items.ToList(); // Return copy
        }
        finally
        {
            _lock.ExitReadLock();
        }
    }
    
    public void Dispose()
    {
        _lock?.Dispose();
    }
}
```

Thread-safe patterns for different scenarios:

```csharp
// Pattern 1: Lazy initialization
public class LazyInitExample
{
    private static readonly Lazy<ExpensiveResource> _resource = 
        new Lazy<ExpensiveResource>(() => new ExpensiveResource());
    
    public static ExpensiveResource Resource => _resource.Value;
}

// Pattern 2: Thread-safe singleton
public sealed class ThreadSafeSingleton
{
    private static readonly object _lock = new object();
    private static ThreadSafeSingleton _instance;
    
    public static ThreadSafeSingleton Instance
    {
        get
        {
            if (_instance == null)
            {
                lock (_lock)
                {
                    if (_instance == null)
                        _instance = new ThreadSafeSingleton();
                }
            }
            return _instance;
        }
    }
}

// Pattern 3: Using concurrent collections
public class ConcurrentDataStore
{
    private readonly ConcurrentDictionary<string, object> _data = 
        new ConcurrentDictionary<string, object>();
    
    public void Store(string key, object value)
    {
        _data.AddOrUpdate(key, value, (k, v) => value);
    }
    
    public T Get<T>(string key)
    {
        return _data.TryGetValue(key, out var value) ? (T)value : default(T);
    }
}
```

## Thread-safe alternatives

| Non-thread-safe | Thread-safe alternative |
|-----------------|------------------------|
| `List<T>` | `ConcurrentBag<T>`, `ImmutableList<T>` |
| `Dictionary<K,V>` | `ConcurrentDictionary<K,V>` |
| `Queue<T>` | `ConcurrentQueue<T>` |
| `Stack<T>` | `ConcurrentStack<T>` |
| `int++` | `Interlocked.Increment(ref int)` |
| `StringBuilder` | Use local instances or synchronize |
| `Random` | `ThreadLocal<Random>` or `Random.Shared` (.NET 6+) |

## Best practices

1. **Prefer immutable data structures** when possible
2. **Use concurrent collections** for shared mutable data
3. **Minimize shared state** between threads
4. **Use atomic operations** for simple shared variables
5. **Lock at the smallest scope** necessary
6. **Avoid nested locks** to prevent deadlocks
7. **Consider using async patterns** instead of blocking synchronization
