# ERP022 - Unobserved exception in a generic exception handler

This analyzer detects generic exception handlers that swallow exceptions without properly observing or handling them.

## Description

The analyzer warns when catch blocks handle exceptions in a generic way (like `catch (Exception)` or bare `catch`) but don't properly observe the exception or provide adequate handling. This can hide important errors and make debugging difficult.

## Code that triggers the analyzer

```csharp
public void ProcessData()
{
    try
    {
        RiskyOperation();
    }
    catch (Exception) // ❌ ERP022
    {
        // Bad: exception is completely swallowed
        return;
    }
}

public void ProcessItems()
{
    try
    {
        foreach (var item in items)
        {
            ProcessItem(item);
        }
    }
    catch // ❌ ERP022
    {
        // Bad: bare catch that swallows everything
        // No way to know what went wrong
    }
}

public bool TryProcess()
{
    try
    {
        DoComplexOperation();
        return true;
    }
    catch (Exception) // ❌ ERP022
    {
        // Bad: swallowing exception and returning false
        // Caller has no idea what failed
        return false;
    }
}
```

## How to fix

Properly observe and handle exceptions:

```csharp
public void ProcessData()
{
    try
    {
        RiskyOperation();
    }
    catch (Exception ex) // ✅ Correct
    {
        // Good: log the exception before handling
        _logger.LogError(ex, "Failed to process data");
        
        // Decide whether to re-throw, return, or handle differently
        throw; // or handle appropriately
    }
}

public void ProcessItems()
{
    try
    {
        foreach (var item in items)
        {
            ProcessItem(item);
        }
    }
    catch (Exception ex) // ✅ Correct
    {
        // Good: observe the exception and provide context
        _logger.LogError(ex, "Failed to process items");
        
        // Handle gracefully or re-throw as appropriate
        throw new ProcessingException("Batch processing failed", ex);
    }
}
```
        DoComplexOperation();
        return true;
    }
    catch (Exception ex)
    {
        // Good: log the exception even in Try methods
        _logger.LogWarning(ex, "Process operation failed");
        return false;
    }
}
```

Better patterns for different scenarios:

```csharp
// Pattern 1: Try methods with proper error handling
public bool TryProcessWithDiagnostics(out string error)
{
    error = null;
    try
    {
        DoComplexOperation();
        return true;
    }
    catch (Exception ex)
    {
        error = ex.Message;
        _logger.LogWarning(ex, "Process operation failed");
        return false;
    }
}

// Pattern 2: Specific exception handling with fallback
public void ProcessWithFallback()
{
    try
    {
        RiskyOperation();
    }
    catch (HttpRequestException ex)
    {
        _logger.LogWarning(ex, "Network request failed, using cache");
        UseCache();
    }
    catch (Exception ex)
    {
        _logger.LogError(ex, "Unexpected error in processing");
        throw; // Re-throw unexpected exceptions
    }
}

// Pattern 3: Graceful degradation with proper logging
public async Task<string> GetDataWithFallback()
{
    try
    {
        return await GetFromPrimarySource();
    }
    catch (Exception ex)
    {
        _logger.LogWarning(ex, "Primary source failed, trying fallback");
        
        try
        {
            return await GetFromFallbackSource();
        }
        catch (Exception fallbackEx)
        {
            _logger.LogError(fallbackEx, "Both primary and fallback sources failed");
            throw new DataUnavailableException("All data sources failed", ex);
        }
    }
}
```

Handling different exception types appropriately:

```csharp
public void ProcessFile(string path)
{
    try
    {
        var content = File.ReadAllText(path);
        ProcessContent(content);
    }
    catch (FileNotFoundException ex)
    {
        // Expected exception - log and handle gracefully
        _logger.LogInformation("File not found: {Path}", path);
        UseDefaultContent();
    }
    catch (UnauthorizedAccessException ex)
    {
        // Security issue - log and throw
        _logger.LogError(ex, "Access denied reading file: {Path}", path);
        throw;
    }
    catch (Exception ex)
    {
        // Unexpected exception - log with full details and throw
        _logger.LogError(ex, "Unexpected error processing file: {Path}", path);
        throw;
    }
}
```

## Guidelines for exception handling

1. **Always observe exceptions**: Log or examine the exception before deciding how to handle it
2. **Provide context**: Include relevant information in logs and wrapped exceptions
3. **Handle specifically**: Catch specific exception types when possible
4. **Don't swallow silently**: Unhandled exceptions should be logged at minimum
5. **Consider the caller**: Think about what information the calling code needs
6. **Use appropriate log levels**: Error for unexpected exceptions, Warning for handled issues
