# ERP021 - Incorrect exception propagation

This analyzer detects incorrect exception propagation patterns where the original stack trace is lost.

## Description

The analyzer warns when exceptions are re-thrown incorrectly using `throw ex;` instead of `throw;`. Using `throw ex;` resets the stack trace, losing valuable debugging information about where the exception originally occurred.

## Code that triggers the analyzer

```csharp
public void ProcessData()
{
    try
    {
        RiskyOperation();
    }
    catch (Exception ex)
    {
        LogError(ex);
        
        // Bad: throws the exception object, losing original stack trace
        throw ex; // ❌ ERP021
    }
}

public async Task ProcessAsync()
{
    try
    {
        await RiskyOperationAsync();
    }
    catch (InvalidOperationException ex)
    {
        // Bad: re-throwing with lost stack trace
        throw ex; // ❌ ERP021
    }
}
```

## How to fix

Use `throw;` without the exception variable to preserve the stack trace:

```csharp
public void ProcessData()
{
    try
    {
        RiskyOperation();
    }
    catch (Exception ex)
    {
        LogError(ex);
        
        // Good: preserves original stack trace
        throw; // ✅ Correct
    }
}

public async Task ProcessAsync()
{
    try
    {
        await RiskyOperationAsync();
    }
    catch (InvalidOperationException ex)
    {
        LogError(ex);
        
        // Good: original stack trace preserved
        throw; // ✅ Correct
    }
}
```

When you need to wrap the exception:

```csharp
public void ProcessData()
{
    try
    {
        RiskyOperation();
    }
    catch (Exception ex)
    {
        // Good: wrap exception while preserving original as inner exception
        throw new ProcessingException("Failed to process data", ex);
    }
}

public void ProcessWithContext()
{
    try
    {
        RiskyOperation();
    }
    catch (Exception ex)
    {
        // Good: add context while preserving original exception
        throw new InvalidOperationException($"Processing failed for item {itemId}", ex);
    }
}
```

When you need to modify the exception:

```csharp
public void ProcessData()
{
    try
    {
        RiskyOperation();
    }
    catch (ArgumentException ex)
    {
        // If you must throw the same exception type with modifications
        var newEx = new ArgumentException($"Processing context: {ex.Message}", ex.ParamName, ex);
        throw newEx;
    }
}
```

Pattern for conditional re-throwing:

```csharp
public void ProcessData()
{
    try
    {
        RiskyOperation();
    }
    catch (Exception ex)
    {
        if (ShouldHandleException(ex))
        {
            HandleException(ex);
            return;
        }
        
        // Re-throw without losing stack trace
        throw;
    }
}
```

## Why stack trace preservation matters

- **Debugging**: Original stack trace shows where the error actually occurred
- **Troubleshooting**: Easier to identify root cause of issues
- **Logging**: Better error logs with complete call stack information
- **Development**: Faster bug fixing with accurate error locations

## Exception patterns to follow

```csharp
// ✓ Good patterns
throw;                                    // Preserve stack trace
throw new CustomException("msg", ex);     // Wrap with inner exception
throw new CustomException("msg") { Data = ex.Data }; // Transfer data

// ✗ Bad patterns  
throw ex;                                 // Loses stack trace
throw new Exception(ex.Message);         // Loses original exception
```
