# ERP041 - EventSource class should be sealed

This analyzer detects EventSource classes that are not sealed, which can cause runtime errors and unexpected behavior.

## Description

The analyzer warns when EventSource classes are not declared as `sealed`. EventSource classes should be sealed to ensure proper ETW (Event Tracing for Windows) functionality and to prevent inheritance issues that can cause runtime errors.

## Code that triggers the analyzer

```csharp
using System.Diagnostics.Tracing;

// Bad: EventSource class is not sealed
[EventSource(Name = "MyCompany-MyProduct-MyEventSource")]
public class MyEventSource : EventSource // ❌ ERP041
{
    public static readonly MyEventSource Log = new MyEventSource();
    
    [Event(1, Level = EventLevel.Informational)]
    public void ApplicationStarted(string applicationName)
    {
        WriteEvent(1, applicationName);
    }
}

// Also bad: inheriting from unsealed EventSource
public class ExtendedEventSource : MyEventSource // ❌ ERP041
{
    [Event(2, Level = EventLevel.Warning)]
    public void CustomEvent(string message)
    {
        WriteEvent(2, message);
    }
}
```

## How to fix

Make EventSource classes sealed:

```csharp
using System.Diagnostics.Tracing;

// Good: EventSource class is sealed
[EventSource(Name = "MyCompany-MyProduct-MyEventSource")]
public sealed class MyEventSource : EventSource // ✅ Correct
{
    public static readonly MyEventSource Log = new MyEventSource();
    
    private MyEventSource() { } // ✅ Correct - Private constructor for singleton pattern
    
    [Event(1, Level = EventLevel.Informational)]
    public void ApplicationStarted(string applicationName) // ✅ Correct
    {
        WriteEvent(1, applicationName);
    }
    
    [Event(2, Level = EventLevel.Warning)]
    public void WarningOccurred(string message) // ✅ Correct
    {
        WriteEvent(2, message);
    }
    
    [Event(3, Level = EventLevel.Error)]
    public void ErrorOccurred(string error, string details) // ✅ Correct
    {
        WriteEvent(3, error, details);
    }
}
```

Proper EventSource pattern with all recommendations:

```csharp
using System.Diagnostics.Tracing;

[EventSource(Name = "MyCompany-MyProduct-Logging")]
public sealed class ApplicationEventSource : EventSource // ✅ Correct
{
    // Singleton instance
    public static readonly ApplicationEventSource Log = new ApplicationEventSource(); // ✅ Correct
    
    // Private constructor to enforce singleton pattern
    private ApplicationEventSource() { } // ✅ Correct
    
    // Event methods with proper attributes
    [Event(1, 
        Level = EventLevel.Informational,
        Message = "Application started: {0}",
        Keywords = Keywords.Startup)]
    public void ApplicationStarted(string applicationName)
    {
        if (IsEnabled())
        {
            WriteEvent(1, applicationName);
        }
    }
    
    [Event(2,
        Level = EventLevel.Warning,
        Message = "Warning in {0}: {1}",
        Keywords = Keywords.General)]
    public void Warning(string component, string message)
    {
        if (IsEnabled(EventLevel.Warning, Keywords.General))
        {
            WriteEvent(2, component, message);
        }
    }
    
    [Event(3,
        Level = EventLevel.Error,
        Message = "Error in {0}: {1}",
        Keywords = Keywords.General)]
    public void Error(string component, string error)
    {
        if (IsEnabled(EventLevel.Error, Keywords.General))
        {
            WriteEvent(3, component, error);
        }
    }
    
    // Event keywords for categorization
    public static class Keywords
    {
        public const EventKeywords Startup = (EventKeywords)1;
        public const EventKeywords General = (EventKeywords)2;
        public const EventKeywords Performance = (EventKeywords)4;
    }
}
```

Usage example:

```csharp
public class MyApplication
{
    public void Start()
    {
        // Using the sealed EventSource
        ApplicationEventSource.Log.ApplicationStarted("MyApp v1.0");
        
        try
        {
            DoWork();
        }
        catch (Exception ex)
        {
            ApplicationEventSource.Log.Error("Application", ex.Message);
            throw;
        }
    }
    
    private void DoWork()
    {
        ApplicationEventSource.Log.Warning("Worker", "Processing large dataset");
        // Work implementation
    }
}
```

If you need multiple event sources:

```csharp
// Create separate sealed EventSource classes for different components
[EventSource(Name = "MyCompany-Authentication")]
public sealed class AuthEventSource : EventSource
{
    public static readonly AuthEventSource Log = new AuthEventSource();
    private AuthEventSource() { }
    
    [Event(1, Level = EventLevel.Informational)]
    public void UserLoggedIn(string userId)
    {
        WriteEvent(1, userId);
    }
}

[EventSource(Name = "MyCompany-DataAccess")]
public sealed class DataEventSource : EventSource
{
    public static readonly DataEventSource Log = new DataEventSource();
    private DataEventSource() { }
    
    [Event(1, Level = EventLevel.Informational)]
    public void QueryExecuted(string query, long durationMs)
    {
        WriteEvent(1, query, durationMs);
    }
}
```

## Why EventSource classes should be sealed

1. **ETW Requirements**: EventSource uses reflection and code generation that can break with inheritance
2. **Runtime Errors**: Derived EventSource classes can cause runtime exceptions
3. **Performance**: Sealed classes allow better optimization
4. **Design Intent**: EventSource classes are designed to be final implementations
5. **Tool Support**: ETW tools and analyzers expect sealed EventSource classes

## Best practices for EventSource

- Always make EventSource classes sealed
- Use singleton pattern with static readonly instance
- Use private constructors
- Check `IsEnabled()` before expensive operations
- Use proper event IDs (unique within the EventSource)
- Provide meaningful event names and messages
- Use keywords for event categorization
- Follow ETW parameter type restrictions
